// zod/src/v4/core/to-json-schema.ts
import type * as core from "../core/index.js";
import type * as JSONSchema from "./json-schema.js";
import { type $ZodRegistry, globalRegistry } from "./registries.js";
import type * as schemas from "./schemas.js";
import type { StandardJSONSchemaV1, StandardSchemaWithJSONProps } from "./standard-schema.js";

export type Processor<T extends schemas.$ZodType = schemas.$ZodType> = (
  schema: T,
  ctx: ToJSONSchemaContext,
  json: JSONSchema.BaseSchema,
  params: ProcessParams
) => void;

export interface JSONSchemaGeneratorParams {
  processors: Record<string, Processor>;
  /** A registry used to look up metadata for each schema. Any schema with an `id` property will be extracted as a $def.
   *  @default globalRegistry */
  metadata?: $ZodRegistry<Record<string, any>>;
  /** The JSON Schema version to target.
   * - `"draft-2020-12"` ŌĆö Default. JSON Schema Draft 2020-12
   * - `"draft-07"` ŌĆö JSON Schema Draft 7
   * - `"draft-04"` ŌĆö JSON Schema Draft 4
   * - `"openapi-3.0"` ŌĆö OpenAPI 3.0 Schema Object */
  target?: "draft-04" | "draft-07" | "draft-2020-12" | "openapi-3.0" | ({} & string) | undefined;
  /** How to handle unrepresentable types.
   * - `"throw"` ŌĆö Default. Unrepresentable types throw an error
   * - `"any"` ŌĆö Unrepresentable types become `{}` */
  unrepresentable?: "throw" | "any";
  /** Arbitrary custom logic that can be used to modify the generated JSON Schema. */
  override?: (ctx: {
    zodSchema: schemas.$ZodTypes;
    jsonSchema: JSONSchema.BaseSchema;
    path: (string | number)[];
  }) => void;
  /** Whether to extract the `"input"` or `"output"` type. Relevant to transforms, defaults, coerced primitives, etc.
   * - `"output"` ŌĆö Default. Convert the output schema.
   * - `"input"` ŌĆö Convert the input schema. */
  io?: "input" | "output";
  cycles?: "ref" | "throw";
  reused?: "ref" | "inline";
  external?:
    | {
        registry: $ZodRegistry<{ id?: string | undefined }>;
        uri?: ((id: string) => string) | undefined;
        defs: Record<string, JSONSchema.BaseSchema>;
      }
    | undefined;
}

/**
 * Parameters for the toJSONSchema function.
 */
export type ToJSONSchemaParams = Omit<JSONSchemaGeneratorParams, "processors" | "external">;

/**
 * Parameters for the toJSONSchema function when passing a registry.
 */
export interface RegistryToJSONSchemaParams extends ToJSONSchemaParams {
  uri?: (id: string) => string;
}

export interface ProcessParams {
  schemaPath: schemas.$ZodType[];
  path: (string | number)[];
}

export interface Seen {
  /** JSON Schema result for this Zod schema */
  schema: JSONSchema.BaseSchema;
  /** A cached version of the schema that doesn't get overwritten during ref resolution */
  def?: JSONSchema.BaseSchema;
  defId?: string | undefined;
  /** Number of times this schema was encountered during traversal */
  count: number;
  /** Cycle path */
  cycle?: (string | number)[] | undefined;
  isParent?: boolean | undefined;
  /** Schema to inherit JSON Schema properties from (set by processor for wrappers) */
  ref?: schemas.$ZodType | null;
  /** JSON Schema property path for this schema */
  path?: (string | number)[] | undefined;
}

export interface ToJSONSchemaContext {
  processors: Record<string, Processor>;
  metadataRegistry: $ZodRegistry<Record<string, any>>;
  target: "draft-04" | "draft-07" | "draft-2020-12" | "openapi-3.0" | ({} & string);
  unrepresentable: "throw" | "any";
  override: (ctx: {
    // must be schemas.$ZodType to prevent recursive type resolution error
    zodSchema: schemas.$ZodType;
    jsonSchema: JSONSchema.BaseSchema;
    path: (string | number)[];
  }) => void;
  io: "input" | "output";
  counter: number;
  seen: Map<schemas.$ZodType, Seen>;
  cycles: "ref" | "throw";
  reused: "ref" | "inline";
  external?:
    | {
        registry: $ZodRegistry<{ id?: string | undefined }>;
        uri?: ((id: string) => string) | undefined;
        defs: Record<string, JSONSchema.BaseSchema>;
      }
    | undefined;
}

// function initializeContext<T extends schemas.$ZodType>(inputs: JSONSchemaGeneratorParams<T>): ToJSONSchemaContext<T> {
//   return {
//     processor: inputs.processor,
//     metadataRegistry: inputs.metadata ?? globalRegistry,
//     target: inputs.target ?? "draft-2020-12",
//     unrepresentable: inputs.unrepresentable ?? "throw",
//   };
// }

export function initializeContext(params: JSONSchemaGeneratorParams): ToJSONSchemaContext {
  // Normalize target: convert old non-hyphenated versions to hyphenated versions
  let target: ToJSONSchemaContext["target"] = params?.target ?? "draft-2020-12";
  if (target === "draft-4") target = "draft-04";
  if (target === "draft-7") target = "draft-07";

  return {
    processors: params.processors ?? {},
    metadataRegistry: params?.metadata ?? globalRegistry,
    target,
    unrepresentable: params?.unrepresentable ?? "throw",
    override: (params?.override as any) ?? (() => {}),
    io: params?.io ?? "output",
    counter: 0,
    seen: new Map(),
    cycles: params?.cycles ?? "ref",
    reused: params?.reused ?? "inline",
    external: params?.external ?? undefined,
  };
}

export function process<T extends schemas.$ZodType>(
  schema: T,
  ctx: ToJSONSchemaContext,
  _params: ProcessParams = { path: [], schemaPath: [] }
): JSONSchema.BaseSchema {
  const def = schema._zod.def as schemas.$ZodTypes["_zod"]["def"];

  // check for schema in seens
  const seen = ctx.seen.get(schema);

  if (seen) {
    seen.count++;

    // check if cycle
    const isCycle = _params.schemaPath.includes(schema);
    if (isCycle) {
      seen.cycle = _params.path;
    }

    return seen.schema;
  }

  // initialize
  const result: Seen = { schema: {}, count: 1, cycle: undefined, path: _params.path };
  ctx.seen.set(schema, result);

  // custom method overrides default behavior
  const overrideSchema = schema._zod.toJSONSchema?.();
  if (overrideSchema) {
    result.schema = overrideSchema as any;
  } else {
    const params = {
      ..._params,
      schemaPath: [..._params.schemaPath, schema],
      path: _params.path,
    };

    if (schema._zod.processJSONSchema) {
      schema._zod.processJSONSchema(ctx, result.schema, params);
    } else {
      const _json = result.schema;
      const processor = ctx.processors[def.type];
      if (!processor) {
        throw new Error(`[toJSONSchema]: Non-representable type encountered: ${def.type}`);
      }
      processor(schema, ctx, _json, params);
    }

    const parent = schema._zod.parent as T;

    if (parent) {
      // Also set ref if processor didn't (for inheritance)
      if (!result.ref) result.ref = parent;
      process(parent, ctx, params);
      ctx.seen.get(parent)!.isParent = true;
    }
  }

  // metadata
  const meta = ctx.metadataRegistry.get(schema);
  if (meta) Object.assign(result.schema, meta);

  if (ctx.io === "input" && isTransforming(schema)) {
    // examples/defaults only apply to output type of pipe
    delete result.schema.examples;
    delete result.schema.default;
  }

  // set prefault as default
  if (ctx.io === "input" && result.schema._prefault) result.schema.default ??= result.schema._prefault;
  delete result.schema._prefault;

  // pulling fresh from ctx.seen in case it was overwritten
  const _result = ctx.seen.get(schema)!;

  return _result.schema;
}

export function extractDefs<T extends schemas.$ZodType>(
  ctx: ToJSONSchemaContext,
  schema: T
  // params: EmitParams
): void {
  // iterate over seen map;
  const root = ctx.seen.get(schema);

  if (!root) throw new Error("Unprocessed schema. This is a bug in Zod.");

  // Track ids to detect duplicates across different schemas
  const idToSchema = new Map<string, schemas.$ZodType>();
  for (const entry of ctx.seen.entries()) {
    const id = ctx.metadataRegistry.get(entry[0])?.id;
    if (id) {
      const existing = idToSchema.get(id);
      if (existing && existing !== entry[0]) {
        throw new Error(
          `Duplicate schema id "${id}" detected during JSON Schema conversion. Two different schemas cannot share the same id when converted together.`
        );
      }
      idToSchema.set(id, entry[0]);
    }
  }

  // returns a ref to the schema
  // defId will be empty if the ref points to an external schema (or #)
  const makeURI = (entry: [schemas.$ZodType<unknown, unknown>, Seen]): { ref: string; defId?: string } => {
    // comparing the seen objects because sometimes
    // multiple schemas map to the same seen object.
    // e.g. lazy

    // external is configured
    const defsSegment = ctx.target === "draft-2020-12" ? "$defs" : "definitions";
    if (ctx.external) {
      const externalId = ctx.external.registry.get(entry[0])?.id; // ?? "__shared";// `__schema${ctx.counter++}`;

      // check if schema is in the external registry
      const uriGenerator = ctx.external.uri ?? ((id: string) => id);
      if (externalId) {
        return { ref: uriGenerator(externalId) };
      }

      // otherwise, add to __shared
      const id: string = entry[1].defId ?? (entry[1].schema.id as string) ?? `schema${ctx.counter++}`;
      entry[1].defId = id; // set defId so it will be reused if needed
      return { defId: id, ref: `${uriGenerator("__shared")}#/${defsSegment}/${id}` };
    }

    if (entry[1] === root) {
      return { ref: "#" };
    }

    // self-contained schema
    const uriPrefix = `#`;
    const defUriPrefix = `${uriPrefix}/${defsSegment}/`;
    const defId = entry[1].schema.id ?? `__schema${ctx.counter++}`;
    return { defId, ref: defUriPrefix + defId };
  };

  // stored cached version in `def` property
  // remove all properties, set $ref
  const extractToDef = (entry: [schemas.$ZodType<unknown, unknown>, Seen]): void => {
    // if the schema is already a reference, do not extract it
    if (entry[1].schema.$ref) {
      return;
    }
    const seen = entry[1];
    const { ref, defId } = makeURI(entry);

    seen.def = { ...seen.schema };
    // defId won't be set if the schema is a reference to an external schema
    // or if the schema is the root schema
    if (defId) seen.defId = defId;
    // wipe away all properties except $ref
    const schema = seen.schema;
    for (const key in schema) {
      delete schema[key];
    }
    schema.$ref = ref;
  };

  // throw on cycles

  // break cycles
  if (ctx.cycles === "throw") {
    for (const entry of ctx.seen.entries()) {
      const seen = entry[1];
      if (seen.cycle) {
        throw new Error(
          "Cycle detected: " +
            `#/${seen.cycle?.join("/")}/<root>` +
            '\n\nSet the `cycles` parameter to `"ref"` to resolve cyclical schemas with defs.'
        );
      }
    }
  }

  // extract schemas into $defs
  for (const entry of ctx.seen.entries()) {
    const seen = entry[1];

    // convert root schema to # $ref
    if (schema === entry[0]) {
      extractToDef(entry); // this has special handling for the root schema
      continue;
    }

    // extract schemas that are in the external registry
    if (ctx.external) {
      const ext = ctx.external.registry.get(entry[0])?.id;
      if (schema !== entry[0] && ext) {
        extractToDef(entry);
        continue;
      }
    }

    // extract schemas with `id` meta
    const id = ctx.metadataRegistry.get(entry[0])?.id;
    if (id) {
      extractToDef(entry);
      continue;
    }

    // break cycles
    if (seen.cycle) {
      // any
      extractToDef(entry);
      continue;
    }

    // extract reused schemas
    if (seen.count > 1) {
      if (ctx.reused === "ref") {
        extractToDef(entry);
        // biome-ignore lint:
        continue;
      }
    }
  }
}

export function finalize<T extends schemas.$ZodType>(
  ctx: ToJSONSchemaContext,
  schema: T
): ZodStandardJSONSchemaPayload<T> {
  const root = ctx.seen.get(schema);
  if (!root) throw new Error("Unprocessed schema. This is a bug in Zod.");

  // flatten refs - inherit properties from parent schemas
  const flattenRef = (zodSchema: schemas.$ZodType) => {
    const seen = ctx.seen.get(zodSchema)!;

    // already processed
    if (seen.ref === null) return;

    const schema = seen.def ?? seen.schema;
    const _cached = { ...schema };

    const ref = seen.ref;
    seen.ref = null; // prevent infinite recursion

    if (ref) {
      flattenRef(ref);

      const refSeen = ctx.seen.get(ref)!;
      const refSchema = refSeen.schema;

      // merge referenced schema into current
      if (refSchema.$ref && (ctx.target === "draft-07" || ctx.target === "draft-04" || ctx.target === "openapi-3.0")) {
        // older drafts can't combine $ref with other properties
        schema.allOf = schema.allOf ?? [];
        schema.allOf.push(refSchema);
      } else {
        Object.assign(schema, refSchema);
      }
      // restore child's own properties (child wins)
      Object.assign(schema, _cached);

      const isParentRef = zodSchema._zod.parent === ref;

      // For parent chain, child is a refinement - remove parent-only properties
      if (isParentRef) {
        for (const key in schema) {
          if (key === "$ref" || key === "allOf") continue;
          if (!(key in _cached)) {
            delete schema[key];
          }
        }
      }

      // When ref was extracted to $defs, remove properties that match the definition
      if (refSchema.$ref && refSeen.def) {
        for (const key in schema) {
          if (key === "$ref" || key === "allOf") continue;
          if (key in refSeen.def && JSON.stringify(schema[key]) === JSON.stringify(refSeen.def[key])) {
            delete schema[key];
          }
        }
      }
    }

    // If parent was extracted (has $ref), propagate $ref to this schema
    // This handles cases like: readonly().meta({id}).describe()
    // where processor sets ref to innerType but parent should be referenced
    const parent = zodSchema._zod.parent;
    if (parent && parent !== ref) {
      // Ensure parent is processed first so its def has inherited properties
      flattenRef(parent);
      const parentSeen = ctx.seen.get(parent);
      if (parentSeen?.schema.$ref) {
        schema.$ref = parentSeen.schema.$ref;
        // De-duplicate with parent's definition
        if (parentSeen.def) {
          for (const key in schema) {
            if (key === "$ref" || key === "allOf") continue;
            if (key in parentSeen.def && JSON.stringify(schema[key]) === JSON.stringify(parentSeen.def[key])) {
              delete schema[key];
            }
          }
        }
      }
    }

    // execute overrides
    ctx.override({
      zodSchema: zodSchema as schemas.$ZodTypes,
      jsonSchema: schema,
      path: seen.path ?? [],
    });
  };

  for (const entry of [...ctx.seen.entries()].reverse()) {
    flattenRef(entry[0]);
  }

  const result: JSONSchema.BaseSchema = {};
  if (ctx.target === "draft-2020-12") {
    result.$schema = "https://json-schema.org/draft/2020-12/schema";
  } else if (ctx.target === "draft-07") {
    result.$schema = "http://json-schema.org/draft-07/schema#";
  } else if (ctx.target === "draft-04") {
    result.$schema = "http://json-schema.org/draft-04/schema#";
  } else if (ctx.target === "openapi-3.0") {
    // OpenAPI 3.0 schema objects should not include a $schema property
  } else {
    // Arbitrary string values are allowed but won't have a $schema property set
  }

  if (ctx.external?.uri) {
    const id = ctx.external.registry.get(schema)?.id;
    if (!id) throw new Error("Schema is missing an `id` property");
    result.$id = ctx.external.uri(id);
  }

  Object.assign(result, root.def ?? root.schema);

  // build defs object
  const defs: JSONSchema.BaseSchema["$defs"] = ctx.external?.defs ?? {};
  for (const entry of ctx.seen.entries()) {
    const seen = entry[1];
    if (seen.def && seen.defId) {
      defs[seen.defId] = seen.def;
    }
  }

  // set definitions in result
  if (ctx.external) {
  } else {
    if (Object.keys(defs).length > 0) {
      if (ctx.target === "draft-2020-12") {
        result.$defs = defs;
      } else {
        result.definitions = defs;
      }
    }
  }

  try {
    // this "finalizes" this schema and ensures all cycles are removed
    // each call to finalize() is functionally independent
    // though the seen map is shared
    const finalized = JSON.parse(JSON.stringify(result));
    Object.defineProperty(finalized, "~standard", {
      value: {
        ...schema["õ$Ś└■▐ĆČY`^Å×DĻ┼mĄÓčņĖE½Æö$aTāŲ─küĢx=×ēWļJ<Ńøtē:ÆN”ė»UŚtåØ═=Ø.ģ═«VųŪŹ▄­b
.ĒßM┐╬t0∙y_ō{Šū└uā■oŃĄüÆ■aL┘.t▓g{-ß¦ČW„Ģ╚>Ó„"l_hņńõ ¾jŽh┤-x9ś9{ØrxA╔ŹÖ─ßns_(šå#S┘E┐▓)z.¦a¼z@∙B®▐╦å åŚc7£╬╚p╝ŹC|!±ūµ#rųo«w8!rn³õ2l.¦EXe£«9óuXJ»«ńt∙╝Ō/,▒Ē│"╠Ź▒¬²3oNW½G ╠uņųTŻ1│4╣|ÄĮg ('}6zAšÓm-§_─ę’2ź®ąm─╠ä"ÕqA
¶tŁÜ®\ÄNßEQĢ³┴;XŖ
s╩HÆ0ł,Ü
ĢģČæE┘^:¾aÅ¤łzqū9cūg²Ćś?▒xĪš¶▄Ķ¹“ECŃ„Ā¢ųźåe+Öo░ė5¼žrf;Ā/H&'\0<C▀Śļ¼
a╗j>Jä§0Õ:▌K9Yh¹č
·ņ &2 ģü<ÉĮXqĶeź ¼ńģT▌ ®/ż*?└ļźY!1┼6ŃÉ¹B¬ęŽćÉŃĮ ¬«±■Æ­¶:§NŃ®ŖĖ@²╔ █³»z:_HUæuN,└J©5J_HU{„`’ģTuÄi┬.¬>ż²ho§=¬^Hš³®└5╔ŁĪs'	┬	▐_r_Hš„6K^`w$Ō¼9cL÷IkN;%7ćōęlŪ²/©*pļ═³p\o Ö5&35N¢]∙”Ŗ%ĶN«GÖŅt1')/óA ŗź·XnüoI╣õ³╦)Ł[╣Ķŗé9üĆMø/©*║ä6eüETrŽŗ¬Ŗ╬z’f]nüI_×5Æ╣įpų┤]ĶøÄ
ĢßEUÕŚ{÷L ąöį¾┬¬ó│n
ßŗO8ēń^gĮ7┤T ā.”AŁĆ“	xqU!1gN┌┌;ó╦«mŃ¬Ü╠{ H▓ Ć╚;r┬U²¼½R»"*ĪŪ╚pų2▀█«×Ū| Xņō ½óīvÅ-`Żś╝	 █ÖSŠ▒Ü;ŃLę
x!M╔}─█P<30┘æw└▌ē╚]0$ŽNr·░ĻŖ°VĮ$é╦łÅ½┤Ā©®ģq■õEVģ¶T”╗÷■ŽŚ
!½R┼EM%g¹ov{ÉY╦ĻÕFŪĻńYę/ZÅCÕ­Ł┴Ü┌©K¶╚q+└49/╝)▄Żīu│%¹f¤Įj¼ńv╗┐%■Fį"┬]ī,┘^¶*n·GT~!ä│Qē╚ē▒č@▌@
"═─=gfÄ“C2┤£┬·!że9’¤k┌š÷C┌═·eĖ╦É¹ß╦ĒŻĒŃüWõ·9KY\äĻ«J®¹╬#┴VŁ÷Gõ¤AÖ:@ęJ┌Q┬#·F¶Aņł¹▒¶n9łFb╝C┌Ņ▒ņēßÕ┘÷█~G╬¦³c║¾─Y®ĖéāZŻĪNIėŹ┬ŗ«ÜķŚé[õ7-ī6	'óØ½:Iģ─ćgsõ­┬½×┌╦#sØ┌+E]ā▐Ģ³.ŲŚh°B¦¼hČwKŚ§«YSa■¤h ö▀DļV▌qĶõėZDŃlĪU®į╚K\└RćÕ▄#ū_Ģŗ²9§¼„īS7õøĶ╣O÷ŗ»Ļī6ń,>dč1Ņę©®v(HØBńÓģŁ█Owü2.nµe2NÖg:ä[$A“÷o└YėÕŅčĒŲ╦└*¾>æ„-“Ū1ŃĒb”═rūE¦▒╦∙žh ^|/­ßżłG~k“dś∙&šPTĖŽ0MŹĆĪ©║oź/é2.Į{3├░#vÖ<A#@aåQż<žmĪųŁ¶Ź╩6ęp┌Ø²Ó"q╔║5Gä┘»³¢`=÷)uo«P.Ŗīä1ĢP╩±”Ą┘)ĒųČZ<§ČZ├ń÷æ°╚═ 9ĒS+ŁźŪ╣fnQĶ║Öq├ž²ņŻ×@║ä┼▓µ“╠ź aä;¹ŚŖõ&ų"2 │┌5BA!ł▀ ▄¬└╦,Ę, #╔ĄłpW„ÄÄ▒x¬)äŌ`═ĪS0é¶ū▌ĆŅ5b╗Eęf░F`Z;³T┐╠S¦Éā┘ĒĪōį¤q±
«b▀ō{å░F,½ćßķ~f1E┤P\*±z┬Š>ś!Z(aMe^[£xĀ«å_öu÷┤└CU=”ŌČp!B°0č!óy`ÖwuŁ§¼#ęų`ŽĮĖ½ņ╣­į■£ ćo[:|cžb¹ū(Ļ«÷ąH§ĆŌäHłŃ2Ił1T)`▐ų╩°6Ė▒0ļLcQ>5 ³N¶{)&Ų║|#by┴W§°¬«wZC®ż|%§“ć└WMö{f%Ą­TxÄ∙Šņ|└█╚^­š▄ĀĻĮÓ½┼E¹Ń»kų┴¹é»Ng^vBOf▒^į║%A/°ĻĢ4GdN_x┼Kdzwäź÷!░Pmŗß/żWÓi/·¬Ś#mc^╠¼ņł}q}5wV └∙ó»¢ŌiĀ╚”xJ£m0a®ė ļ^²¤╣©ĪU╠│
;ZŠ­½TĆŲ!ė6°*ŻNµĖ5(ģac▐°Ü	2Æż¦∙č{└¤┤- ░3Č/³ĻO³Ułæ~ X│@æ/³j¾łĶĶŲ¬Gm”²ģ
ĄT¦ÓżĒćQĢ V■Ī§źČ«„═ģÄN¹ŗ┐ÜėxŲŖ▄„ķ╣R I├ŗ┐╩lg-¦”ōµō«ėßZ=M▌T)▓f┬ŁąšeĶ∙čÄõBÖŠŹ─▌0®ā@¬eę(u¾r┐ÓsSłōų=nīóŁZńģ┐ać§1rł22ęÜČqSŠ&ā¼m▒═·┴7A}
įĄÕazzyÆ▒B`}'MŻRxP+^ų+Ņ─ĶV¬|ą8 ļĪW^`āĄĖ▐Nu┌S¦┘(┬╬┴/kRWÕóĘ^8mńfÖgä«¤dļŁŚ„ÕÜNmg=═Ś┴ļn@╦ŚŪgÅ·┬<'
∙┴Ķ∙	_ŖlÓÉ /k©©_▌v-ÜŚÕ{§õXe▀Ł∙¼Õüļż╣ć”NVÜ.”■ŗQ|čJ╣x]tŃJ╬l╗ūĶ:.÷ß┌ų│÷¢╦Č×┤¤▌)▐ßR9Ė$ķW├Uī▌#ø¾├ju@¹±÷>ä┬šqėe{¶¤¶\ŁĪ╗%¶w┌$ź┬ŚC├·EaŹ1ąŻĪGūE×čQ╗"ć/Ē“}QX}_0UpA<▀▄Fq®Ö╝rĀ─hS8t¹84ĀŚ©nEä§ŲĆ╣“čć▒∙ń╩_<Ą² ļ▓;Ķ╝`ŗļéü┘¶Eų─*╬tYŖØ¬┤Ź÷ĪyČÉ>4|m▒
Ō!█b-³{Ć„ŗ°ģ┤ņŌ@ró¶┬zVTx╗~·żXßx▄Ų/F┼cł§p8Ś#║O▄`@┴ $':ż”sv}h╔▄Nł╔▐T /bū§sõ¢d'Ō╗C­öĀLgō¼╚ dxvń2Ųmf/│¬ėMdu▐ kŪ\ł╩f─2ÄE╝H¼▀Ż­B▒┬+~<q┴2éb%źg°½ĀæeG▐┼¬ó!4zęOA╗Ćb5F╣aFĄørdā·0└#jŁī~ļ“@t░\>Ó~▓:┼zR»─Ļ±┘ŻŠ\MÅčĒ┘]e f½æ×²/─ĆFóĶ!+Ļ┼bmYõN¾üJ’ßFw└gx¢Ū	|ū1¢w9x1<į┐X¼┐cåŹ
·¢§é▒óÕÖwĪ$sPk┌*Ē╬Śu~┴X!żó³l„Ć~H%Źī
l+ÄRA<Įh¼ąĖ▀c„╝U~čX#Ö¹C]M]LV█ås┤d╠ó▒Žn│h╠b█7m_j ¼ā┼÷ŽlĄįj╔žŚZ]▓Žó└ŠNMHR¹7Ł▀,“Čh╠RSŚ„L∙°├ Ź²$ßO╗E
¢┴X_¹╔Ŗ"ļ1 Å└┘*Ź,kF)Č>▀ ,┘Bc
CW5×ÜNŁ'=║=m=2NzNåOå/:ü''▌ōcĢ╚²''*ē+ćną┤╩ź¦ĒLn░[3¼ō├─│ułōĶĘóź¦Ą¦Łńp='├'²ÅżĀ\ŲX7Ś3Ģ╦2:'ĒgcW╬õ]!iEū╔žŻ^3Õ+ļŲöįĆī&m0¼9¬ū ŃŲp>ļŻė^Ņł*>Žrø&└╚!l_}7dŖžYĪGČģGg@Ś2┘°öł ─īwšŗŪ¬”ķ"k²▀÷źŲ»©¹·¼Ēé,Ō.C]Ł_9Xj¹åHjĄhĶ½Ļ>ė IŁøhČ©╦bļĄ]4Ŗ║┤[4³»═ó■Ž1³┘äōd°½Kå>ćĻ─9vč-Čļ¤¦ń½ĒtōjÓ┼cšp^ų¤vKŁ?█Š░§3█ŠwÄ`▐¤6ŗŲ>Č▒ÅżĒ=<ī““7ū■AG∙Ęł<─Ų¢▄B▐ŽÆ«╝┐Yµ3┐═³Wb!ļ_Ķ┘ÕīfX žRŲ\═3į█ÓE“ß ÆĪ³ÉyÉ`É³╗SĆPĮ{3╗ró­X┐·¢ą]{³IGw÷)½'ŲŖ║/5mS╦×*╗Aī¼ā┴└ø─šś╬ļŚ'tP]ø©”īÄ▄Č;l·½}d#ßī£)= ŃÓżČ┴f┤qÉ52.╝ī»õuēĻ*ķ╩²4/ +ä┐ö X ┼}╩▀┬─ŹģDØ¶ÉuW╚× ł╩k%]¶
