// @babel/traverse/lib/path/evaluation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.evaluate = evaluate;
exports.evaluateTruthy = evaluateTruthy;
const VALID_OBJECT_CALLEES = ["Number", "String", "Math"];
const VALID_IDENTIFIER_CALLEES = ["isFinite", "isNaN", "parseFloat", "parseInt", "decodeURI", "decodeURIComponent", "encodeURI", "encodeURIComponent", null, null];
const INVALID_METHODS = ["random"];
function isValidObjectCallee(val) {
  return VALID_OBJECT_CALLEES.includes(val);
}
function isValidIdentifierCallee(val) {
  return VALID_IDENTIFIER_CALLEES.includes(val);
}
function isInvalidMethod(val) {
  return INVALID_METHODS.includes(val);
}
function evaluateTruthy() {
  const res = this.evaluate();
  if (res.confident) return !!res.value;
}
function deopt(path, state) {
  if (!state.confident) return;
  state.deoptPath = path;
  state.confident = false;
}
const Globals = new Map([["undefined", undefined], ["Infinity", Infinity], ["NaN", NaN]]);
function evaluateCached(path, state) {
  const {
    node
  } = path;
  const {
    seen
  } = state;
  if (seen.has(node)) {
    const existing = seen.get(node);
    if (existing.resolved) {
      return existing.value;
    } else {
      deopt(path, state);
      return;
    }
  } else {
    const item = {
      resolved: false
    };
    seen.set(node, item);
    const val = _evaluate(path, state);
    if (state.confident) {
      item.resolved = true;
      item.value = val;
    }
    return val;
  }
}
function _evaluate(path, state) {
  if (!state.confident) return;
  if (path.isSequenceExpression()) {
    const exprs = path.get("expressions");
    return evaluateCached(exprs[exprs.length - 1], state);
  }
  if (path.isStringLiteral() || path.isNumericLiteral() || path.isBooleanLiteral()) {
    return path.node.value;
  }
  if (path.isNullLiteral()) {
    return null;
  }
  if (path.isTemplateLiteral()) {
    return evaluateQuasis(path, path.node.quasis, state);
  }
  if (path.isTaggedTemplateExpression() && path.get("tag").isMemberExpression()) {
    const object = path.get("tag.object");
    const {
      node: {
        name
      }
    } = object;
    const property = path.get("tag.property");
    if (object.isIdentifier() && name === "String" && !path.scope.getBinding(name) && property.isIdentifier() && property.node.name === "raw") {
      return evaluateQuasis(path, path.node.quasi.quasis, state, true);
    }
  }
  if (path.isConditionalExpression()) {
    const testResult = evaluateCached(path.get("test"), state);
    if (!state.confident) return;
    if (testResult) {
      return evaluateCached(path.get("consequent"), state);
    } else {
      return evaluateCached(path.get("alternate"), state);
    }
  }
  if (path.isExpressionWrapper()) {
    return evaluateCached(path.get("expression"), state);
  }
  if (path.isMemberExpression() && !path.parentPath.isCallExpression({
    callee: path.node
  })) {
    const property = path.get("property");
    const object = path.get("object");
    if (object.isLiteral()) {
      const value = object.node.value;
      const type = typeof value;
      let key = null;
      if (path.node.computed) {
        key = evaluateCached(property, state);
        if (!state.confident) return;
      } else if (property.isIdentifier()) {
        key = property.node.name;
      }
      if ((type === "number" || type === "string") && key != null && (typeof key === "number" || typeof key === "string")) {
        return value[key];
      }
    }
  }
  if (path.isReferencedIdentifier()) {
    const binding = path.scope.getBinding(path.node.name);
    if (binding) {
      if (binding.constantViolations.length > 0 || path.node.start < binding.path.node.end) {
        deopt(binding.path, state);
        return;
      }
      const bindingPathScope = binding.path.scope;
      if (binding.kind === "var" && bindingPathScope !== binding.scope) {
        let hasUnsafeBlock = !bindingPathScope.path.parentPath.isBlockStatement();
        for (let scope = bindingPathScope.parent; scope; scope = scope.parent) {
          var _scope$path$parentPat;
   value:n});const p=g.concat(c),m=(0,o.x)(p),f=(0,n.w)(p);var v,y;t&&(null===(v=document)||void 0===v||null===(y=v.body)||void 0===y||y.style.setProperty("background-color",Ae.kc));return{stylingProvider:h,initialTheme:m,initialThemeV9:f,theming:u}}async function Je(e,t){return await(0,De.RG)(new URL(e)).catch(e=>((0,Ne.vV)(`Unable to get fluid file name from Sharepoint.\n${(0,Pe.u1)(e)}`),null))||function(e){try{const t=/^.*[/\\]([^/\\]+)$/,r=new URL(e).pathname,o=t.exec(r);if(o&&o.length>1){const e=decodeURIComponent(o[1]);if(e.endsWith(we.g4)||e.endsWith(we.Ct))return e}}catch(t){}return null}(e)||t||""}function Ke(e,t){const r=document.documentElement;Ue=r?new Le(r,e,!1,Ie.CQ,t):void 0}function Qe(e){switch(e){case"Automation":case"Local":case"Staging":case"SyntheticTraffic":return{audience:"Staging"};case"Dogfood":return{audience:"Dogfood"};case"Microsoft":return{audience:"Microsoft"};case"Preview":return{audience:"Preview"};case"DoD":return{audience:"DoD"};case"GCC":return{audience:"GCC"};case"GCC-High":return{audience:"GCC-High"};case"Gallatin":return{audience:"Gallatin"};case"Bleu":return{audience:"Bleu"};case"Delos":return{audience:"Delos"};case"USNat":return{audience:"USNat"};case"USSec":return{audience:"USSec"};case"Production":return{audience:"Production"};default:return we.Nu||(0,Ne.vV)(`Audience ${e} is not a known value.`),{audience:"Production"}}}function Ye(e){switch(e){case"Demo App":case"Demo Host":case"Fluid Preview App":case"Loop App":case"Office.com":case"OfficeMobile":case"OneNote":case"Outlook":case"Scriptor Pad":case"Teams":case"Whiteboard":case"Word":return e;default:return we.Nu||(0,Ne.Rm)(`"${e}" is not a valid app name, set to Unknown`),"Unknown"}}function Ze(e,t,r,o){const n=(0,Te.xE)(),i=function(e){switch(null===e||void 0===e?void 0:e.toLowerCase()){case"android":return"Android";case"ios":return"iOS";case"mac":return"Mac";case"web":return"Web";case"win32":return"Win32";default:throw new Error(`"${e}" is not a valid app platform.`)}}(n.platform),s=Ye(n.appName);return(0,Ne.Rm)(`Fluid host context: {\n  correlationId: ${t}, \n  hostAppName: ${s} \n  hostAppPlatform: ${i} \n  hostAppVersion: ${n.appVersion}, \n  scenarioLifecycle: ${o.lifecycle}, \n  scenarioName: ${o.name}, \n  scenarioEntryPoint: ${o.entryPoint}, \n}`),{hostName:`${n.appName}_${i}`,hostAppName:s,hostAppPlatform:i,hostSessionId:e,hostCorrelationId:t,uiLanguage:r,hostAppVersion:n.appVersion,hostScenarioEntryPoint:o.entryPoint,hostScenarioLifecycle:o.lifecycle,hostScenarioName:o.name,hostScenarioStartTime:o.startTime,userContext:{}}}},90589:(e,t,r)=>{"use strict";function o(e,t,r){return a(e,1,t,r)}function n(e,t,r){return a(e,2,t,r)}function i(e,t,r){return a(e,3,t,r)}function s(e,t,r){return a(e,0,t,r)}function a(e,t,r,o){return{name:e,dataType:t,value:r,classification:o||4}}r.d(t,{LK:()=>o,Ug:()=>n,ip:()=>i,m9:()=>s})},90662:(e,t,r)=>{"use strict";r.d(t,{kc:()=>d,h2:()=>g,mz:()=>s,d8:()=>c,D8:()=>a,g4:()=>h,fB:()=>p});r(899);var o=r(90140);let n=function(e){return e[e.Default=0]="Default",e[e.Dark=1]="Dark",e}({});const i=new Map([[n[n.Default],{name:"Default Theme",properties:[]}],[n[n.Dark],{name:"Dark Theme",properties:[[o.Ve.black,"#ffffff"],[o.Ve.white,"#262626"],[o.Ve.neutralDark,"#ffffff"],[o.Ve.neutralPrimary,"#ffffff"],[o.Ve.neutralPrimaryAlt,"#f8f8f8"],[o.Ve.neutralSecondary,"#f4f4f4"],[o.Ve.neutralTertiary,"#eaeaea"],[o.Ve.neutralTertiaryAlt,"#c8c8c8"],[o.Ve.neutralQuaternary,"#2e2e2e"],[o.Ve.neutralLight,"#000000"],[o.Ve.neutralQuaternaryAlt,"#2e2e2e"],[o.Ve.neutralLighter,"#212121"],[o.Ve.neutralLighterAlt,"#363636"],[o.Ve.redDark,"#e81123"],[o.Ve.purple,"#9c89e9"],[o.Ve.themeDarker,"#cfd3d5"],[o.Ve.themeDark,"#bcd6e6"],[o.Ve.themeDarkAlt,"#afd8f0"],[o.Ve.themePrimary,"#79b0c8"],[o.Ve.themeSecondary,"#588b9e"],[o.Ve.themeTertiary,"#597c87"],[o.Ve.themeLight,"#495e66"],[o.Ve.themeLighter,"#485b60"],[o.Ve.themeLighterAlt,"#3c4547"],[o.Ve.EntityRecognitionUnderline,"#8a8886"],[o.Ve.EntityRecognitionHoverHighlight,"#3b3a39"],[o.Ve.TaskRecognitionUnderline,"#8a8886"],[o.Ve.TaskRecognitionHoverHighlight,"#3b3a39"],[o._        return left / right;
      case "*":
        return left * right;
      case "%":
        return left % right;
      case "**":
        return Math.pow(left, right);
      case "<":
        return left < right;
      case ">":
        return left > right;
      case "<=":
        return left <= right;
      case ">=":
        return left >= right;
      case "==":
        return left == right;
      case "!=":
        return left != right;
      case "===":
        return left === right;
      case "!==":
        return left !== right;
      case "|":
        return left | right;
      case "&":
        return left & right;
      case "^":
        return left ^ right;
      case "<<":
        return left << right;
      case ">>":
        return left >> right;
      case ">>>":
        return left >>> right;
    }
  }
  if (path.isCallExpression()) {
    const callee = path.get("callee");
    let context;
    let func;
    if (callee.isIdentifier() && !path.scope.getBinding(callee.node.name) && (isValidObjectCallee(callee.node.name) || isValidIdentifierCallee(callee.node.name))) {
      func = global[callee.node.name];
    }
    if (callee.isMemberExpression()) {
      const object = callee.get("object");
      const property = callee.get("property");
      if (object.isIdentifier() && property.isIdentifier() && isValidObjectCallee(object.node.name) && !isInvalidMethod(property.node.name)) {
        context = global[object.node.name];
        const key = property.node.name;
        if (hasOwnProperty.call(context, key)) {
          func = context[key];
        }
      }
      if (object.isLiteral() && property.isIdentifier()) {
        const type = typeof object.node.value;
        if (type === "string" || type === "number") {
          context = object.node.value;
          func = context[property.node.name];
        }
      }
    }
    if (func) {
      const args = path.get("arguments").map(arg => evaluateCached(arg, state));
      if (!state.confident) return;
      return func.apply(context, args);
    }
  }
  deopt(path, state);
}
function evaluateQuasis(path, quasis, state, raw = false) {
  let str = "";
  let i = 0;
  const exprs = path.isTemplateLiteral() ? path.get("expressions") : path.get("quasi.expressions");
  for (const elem of quasis) {
    if (!state.confident) break;
    str += raw ? elem.value.raw : elem.value.cooked;
    const expr = exprs[i++];
    if (expr) str += String(evaluateCached(expr, state));
  }
  if (!state.confident) return;
  return str;
}
function evaluate() {
  const state = {
    confident: true,
    deoptPath: null,
    seen: new Map()
  };
  let value = evaluateCached(this, state);
  if (!state.confident) value = undefined;
  return {
    confident: state.confident,
    deopt: state.deoptPath,
    value: value
  };
}
