// zod/v4/mini/schemas.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZodMiniFile = exports.ZodMiniLiteral = exports.ZodMiniEnum = exports.ZodMiniSet = exports.ZodMiniMap = exports.ZodMiniRecord = exports.ZodMiniTuple = exports.ZodMiniIntersection = exports.ZodMiniDiscriminatedUnion = exports.ZodMiniXor = exports.ZodMiniUnion = exports.ZodMiniObject = exports.ZodMiniArray = exports.ZodMiniDate = exports.ZodMiniVoid = exports.ZodMiniNever = exports.ZodMiniUnknown = exports.ZodMiniAny = exports.ZodMiniNull = exports.ZodMiniUndefined = exports.ZodMiniSymbol = exports.ZodMiniBigIntFormat = exports.ZodMiniBigInt = exports.ZodMiniBoolean = exports.ZodMiniNumberFormat = exports.ZodMiniNumber = exports.ZodMiniCustomStringFormat = exports.ZodMiniJWT = exports.ZodMiniE164 = exports.ZodMiniBase64URL = exports.ZodMiniBase64 = exports.ZodMiniMAC = exports.ZodMiniCIDRv6 = exports.ZodMiniCIDRv4 = exports.ZodMiniIPv6 = exports.ZodMiniIPv4 = exports.ZodMiniKSUID = exports.ZodMiniXID = exports.ZodMiniULID = exports.ZodMiniCUID2 = exports.ZodMiniCUID = exports.ZodMiniNanoID = exports.ZodMiniEmoji = exports.ZodMiniURL = exports.ZodMiniUUID = exports.ZodMiniGUID = exports.ZodMiniEmail = exports.ZodMiniStringFormat = exports.ZodMiniString = exports.ZodMiniType = void 0;
exports.ZodMiniFunction = exports.stringbool = exports.meta = exports.describe = exports.ZodMiniCustom = exports.ZodMiniPromise = exports.ZodMiniLazy = exports.ZodMiniTemplateLiteral = exports.ZodMiniReadonly = exports.ZodMiniCodec = exports.ZodMiniPipe = exports.ZodMiniNaN = exports.ZodMiniCatch = exports.ZodMiniSuccess = exports.ZodMiniNonOptional = exports.ZodMiniPrefault = exports.ZodMiniDefault = exports.ZodMiniNullable = exports.ZodMiniExactOptional = exports.ZodMiniOptional = exports.ZodMiniTransform = void 0;
exports.string = string;
exports.email = email;
exports.guid = guid;
exports.uuid = uuid;
exports.uuidv4 = uuidv4;
exports.uuidv6 = uuidv6;
exports.uuidv7 = uuidv7;
exports.url = url;
exports.httpUrl = httpUrl;
exports.emoji = emoji;
exports.nanoid = nanoid;
exports.cuid = cuid;
exports.cuid2 = cuid2;
exports.ulid = ulid;
exports.xid = xid;
exports.ksuid = ksuid;
exports.ipv4 = ipv4;
exports.ipv6 = ipv6;
exports.cidrv4 = cidrv4;
exports.cidrv6 = cidrv6;
exports.mac = mac;
exports.base64 = base64;
exports.base64url = base64url;
exports.e164 = e164;
exports.jwt = jwt;
exports.stringFormat = stringFormat;
exports.hostname = hostname;
exports.hex = hex;
exports.hash = hash;
exports.number = number;
exports.int = int;
exports.float32 = float32;
exports.float64 = float64;
exports.int32 = int32;
exports.uint32 = uint32;
exports.boolean = boolean;
exports.bigint = bigint;
exports.int64 = int64;
exports.uint64 = uint64;
exports.symbol = symbol;
exports.undefined = _undefined;
exports.null = _null;
exports.any = any;
exports.unknown = unknown;
exports.never = never;
exports.void = _void;
exports.date = date;
exports.array = array;
exports.keyof = keyof;
exports.object = object;
exports.strictObject = strictObject;
exports.looseObject = looseObject;
exports.extend = extend;
exports.safeExtend = safeExtend;
exports.merge = merge;
exports.pick = pick;
exports.omit = omit;
exports.partial = partial;
exports.required = required;
exports.catchall = catchall;
exports.union = union;
exports.xor = xor;
exports.discriminatedUnion = discriminatedUnion;
exports.intersection = intersection;
exports.tuple = tuple;
exports.record = record;
exports.partialRecord = partialRecord;
exports.looseRecord = looseRecord;
exports.map = map;
exports.set = set;
exports.enum = _enum;
exports.nativeEnum = nativeEnum;
exports.literal = literal;
exports.file = file;
exports.transform = transform;
exports.optional = optional;
exports.exactOptional = exactOptional;
exports.nullable = nullable;
exports.nullish = nullish;
exports._default = _default;
exports.prefault = prefault;
exports.nonoptional = nonoptional;
exports.success = success;
exports.catch = _catch;
exports.nan = nan;
exports.pipe = pipe;
exports.codec = codec;
exports.readonly = readonly;
exports.templateLiteral = templateLiteral;
exports.lazy = _lazy;
exports.promise = promise;
exports.check = check;
exports.custom = custom;
exports.refine = refine;
exports.superRefine = superRefine;
exports.instanceof = _instanceof;
exports.json = json;
exports._function = _function;
exports.function = _function;
exports._function = _function;
exports.function = _function;
const core = __importStar(require("../core/index.cjs"));
const util = __importStar(require("../core/util.cjs"));
const parse = __importStar(require("./parse.cjs"));
exports.ZodMiniType = core.$constructor("ZodMiniType", (inst, def) => {
    if (!inst._zod)
        throw new Error("Uninitialized schema in ZodMiniType.");
    core.$ZodType.init(inst, def);
    inst.def = def;
    inst.type = def.type;
    inst.parse = (data, params) => parse.parse(inst, data, params, { callee: inst.parse });
    inst.safeParse = (data, params) => parse.safeParse(inst, data, params);
    inst.parseAsync = async (data, params) => parse.parseAsync(inst, data, params, { callee: inst.parseAsync });
    inst.safeParseAsync = async (data, params) => parse.safeParseAsync(inst, data, params);
    inst.check = (...checks) => {
        return inst.clone({
            ...def,
            checks: [
                ...(def.checks ?? []),
                ...checks.map((ch) => typeof ch === "function" ? { _zod: { check: ch, def: { check: "custom" }, onattach: [] } } : ch),
            ],
        }, { parent: true });
    };
    inst.with = inst.check;
    inst.clone = (_def, params) => core.clone(inst, _def, params);
    inst.brand = () => inst;
    inst.register = ((reg, meta) => {
        reg.add(inst, meta);
        return inst;
    });
    inst.apply = (fn) => fn(inst);
});
exports.ZodMiniString = core.$constructor("ZodMiniString", (inst, def) => {
    core.$ZodString.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function string(params) {
    return core._string(exports.ZodMiniString, params);
}
exports.ZodMiniStringFormat = core.$constructor("ZodMiniStringFormat", (inst, def) => {
    core.$ZodStringFormat.init(inst, def);
    exports.ZodMiniString.init(inst, def);
});
exports.ZodMiniEmail = core.$constructor("ZodMiniEmail", (inst, def) => {
    core.$ZodEmail.init(inst, def);
    exports.ZodMiniStringFormat.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function email(params) {
    return core._email(exports.ZodMiniEmail, params);
}
exports.ZodMiniGUID = core.$constructor("ZodMiniGUID", (inst, def) => {
    core.$ZodGUID.init(inst, def);
    exports.ZodMiniStringFormat.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function guid(params) {
    return core._guid(exports.ZodMiniGUID, params);
}
exports.ZodMiniUUID = core.$constructor("ZodMiniUUID", (inst, def) => {
    core.$ZodUUID.init(inst, def);
    exports.ZodMiniStringFormat.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function uuid(params) {
    return core._uuid(exports.ZodMiniUUID, params);
}
// @__NO_SIDE_EFFECTS__
function uuidv4(params) {
    return core._uuidv4(exports.ZodMiniUUID, params);
}
// ZodMiniUUIDv6
// @__NO_SIDE_EFFECTS__
function uuidv6(params) {
    return core._uuidv6(exports.ZodMiniUUID, params);
}
// ZodMiniUUIDv7
// @__NO_SIDE_EFFECTS__
function uuidv7(params) {
    return core._uuidv7(exports.ZodMiniUUID, params);
}
exports.ZodMiniURL = core.$constructor("ZodMiniURL", (inst, def) => {
    core.$ZodURL.init(inst, def);
    exports.ZodMiniStringFormat.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function url(params) {
    return core._url(exports.ZodMiniURL, params);
}
// @__NO_SIDE_EFFECTS__
function httpUrl(params) {
    return core._url(exports.ZodMiniURL, {
        protocol: /^https?$/,
        hostname: core.regexes.domain,
        ...util.normalizeParams(params),
    });
}
exports.ZodMiniEmoji = core.$constructor("ZodMiniEmoji", (inst, def) => {
    core.$ZodEmoji.init(inst, def);
    exports.ZodMiniStringFormat.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function emoji(params) {
    return core._emoji(exports.ZodMiniEmoji, params);
}
exports.ZodMiniNanoID = core.$constructor("ZodMiniNanoID", (inst, def) => {
    core.$ZodNanoID.init(inst, def);
    exports.ZodMiniStringFormat.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function nanoid(params) {
    return core._nanoid(exports.ZodMiniNanoID, params);
}
exports.ZodMiniCUID = core.$constructor("ZodMiniCUID", (inst, def) => {
    core.$ZodCUID.init(inst, def);
    exports.ZodMiniStringFormat.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function cuid(params) {
    return core._cuid(exports.ZodMiniCUID, params);
}
exports.ZodMiniCUID2 = core.$constructor("ZodMiniCUID2", (inst, def) => {
    core.$ZodCUID2.init(inst, def);
    exports.ZodMiniStringFormat.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function cuid2(params) {
    return core._cuid2(exports.ZodMiniCUID2, params);
}
exports.ZodMiniULID = core.$constructor("ZodMiniULID", (inst, def) => {
    core.$ZodULID.init(inst, def);
    exports.ZodMiniStringFormat.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function ulid(params) {
    return core._ulid(exports.ZodMiniULID, params);
}
exports.ZodMiniXID = core.$constructor("ZodMiniXID", (inst, def) => {
    core.$ZodXID.init(inst, def);
    exports.ZodMiniStringFormat.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function xid(params) {
    return core._xid(exports.ZodMiniXID, params);
}
exports.ZodMiniKSUID = core.$constructor("ZodMiniKSUID", (inst, def) => {
    core.$ZodKSUID.init(inst, def);
    exports.ZodMiniStringFormat.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function ksuid(params) {
    return core._ksuid(exports.ZodMiniKSUID, params);
}
exports.ZodMiniIPv4 = core.$constructor("ZodMiniIPv4", (inst, def) => {
    core.$ZodIPv4.init(inst, def);
    exports.ZodMiniStringFormat.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function ipv4(params) {
    return core._ipv4(exports.ZodMiniIPv4, params);
}
exports.ZodMiniIPv6 = core.$constructor("ZodMiniIPv6", (inst, def) => {
    core.$ZodIPv6.init(inst, def);
    exports.ZodMiniStringFormat.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function ipv6(params) {
    return core._ipv6(exports.ZodMiniIPv6, params);
}
exports.ZodMiniCIDRv4 = core.$constructor("ZodMiniCIDRv4", (inst, def) => {
    core.$ZodCIDRv4.init(inst, def);
    exports.ZodMiniStringFormat.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function cidrv4(params) {
    return core._cidrv4(exports.ZodMiniCIDRv4, params);
}
exports.ZodMiniCIDRv6 = core.$constructor("ZodMiniCIDRv6", (inst, def) => {
    core.$ZodCIDRv6.init(inst, def);
    exports.ZodMiniStringFormat.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function cidrv6(params) {
    return core._cidrv6(exports.ZodMiniCIDRv6, params);
}
exports.ZodMiniMAC = core.$constructor("ZodMiniMAC", (inst, def) => {
    core.$ZodMAC.init(inst, def);
    exports.ZodMiniStringFormat.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function mac(params) {
    return core._mac(exports.ZodMiniMAC, params);
}
exports.ZodMiniBase64 = core.$constructor("ZodMiniBase64", (inst, def) => {
    core.$ZodBase64.init(inst, def);
    exports.ZodMiniStringFormat.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function base64(params) {
    return core._base64(exports.ZodMiniBase64, params);
}
exports.ZodMiniBase64URL = core.$constructor("ZodMiniBase64URL", (inst, def) => {
    core.$ZodBase64URL.init(inst, def);
    exports.ZodMiniStringFormat.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function base64url(params) {
    return core._base64url(exports.ZodMiniBase64URL, params);
}
exports.ZodMiniE164 = core.$constructor("ZodMiniE164", (inst, def) => {
    core.$ZodE164.init(inst, def);
    exports.ZodMiniStringFormat.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function e164(params) {
    return core._e164(exports.ZodMiniE164, params);
}
exports.ZodMiniJWT = core.$constructor("ZodMiniJWT", (inst, def) => {
    core.$ZodJWT.init(inst, def);
    exports.ZodMiniStringFormat.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function jwt(params) {
    return core._jwt(exports.ZodMiniJWT, params);
}
exports.ZodMiniCustomStringFormat = core.$constructor("ZodMiniCustomStringFormat", (inst, def) => {
    core.$ZodCustomStringFormat.init(inst, def);
    exports.ZodMiniStringFormat.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function stringFormat(format, fnOrRegex, _params = {}) {
    return core._stringFormat(exports.ZodMiniCustomStringFormat, format, fnOrRegex, _params);
}
// @__NO_SIDE_EFFECTS__
function hostname(_params) {
    return core._stringFormat(exports.ZodMiniCustomStringFormat, "hostname", core.regexes.hostname, _params);
}
// @__NO_SIDE_EFFECTS__
function hex(_params) {
    return core._stringFormat(exports.ZodMiniCustomStringFormat, "hex", core.regexes.hex, _params);
}
// @__NO_SIDE_EFFECTS__
function hash(alg, params) {
    const enc = params?.enc ?? "hex";
    const format = `${alg}_${enc}`;
    const regex = core.regexes[format];
    // check for unrecognized format
    if (!regex)
        throw new Error(`Unrecognized hash format: ${format}`);
    return core._stringFormat(exports.ZodMiniCustomStringFormat, format, regex, params);
}
exports.ZodMiniNumber = core.$constructor("ZodMiniNumber", (inst, def) => {
    core.$ZodNumber.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function number(params) {
    return core._number(exports.ZodMiniNumber, params);
}
exports.ZodMiniNumberFormat = core.$constructor("ZodMiniNumberFormat", (inst, def) => {
    core.$ZodNumberFormat.init(inst, def);
    exports.ZodMiniNumber.init(inst, def);
});
// int
// @__NO_SIDE_EFFECTS__
function int(params) {
    return core._int(exports.ZodMiniNumberFormat, params);
}
// float32
// @__NO_SIDE_EFFECTS__
function float32(params) {
    return core._float32(exports.ZodMiniNumberFormat, params);
}
// float64
// @__NO_SIDE_EFFECTS__
function float64(params) {
    return core._float64(exports.ZodMiniNumberFormat, params);
}
// int32
// @__NO_SIDE_EFFECTS__
function int32(params) {
    return core._int32(exports.ZodMiniNumberFormat, params);
}
// uint32
// @__NO_SIDE_EFFECTS__
function uint32(params) {
    return core._uint32(exports.ZodMiniNumberFormat, params);
}
exports.ZodMiniBoolean = core.$constructor("ZodMiniBoolean", (inst, def) => {
    core.$ZodBoolean.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function boolean(params) {
    return core._boolean(exports.ZodMiniBoolean, params);
}
exports.ZodMiniBigInt = core.$constructor("ZodMiniBigInt", (inst, def) => {
    core.$ZodBigInt.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function bigint(params) {
    return core._bigint(exports.ZodMiniBigInt, params);
}
exports.ZodMiniBigIntFormat = core.$constructor("ZodMiniBigIntFormat", (inst, def) => {
    core.$ZodBigIntFormat.init(inst, def);
    exports.ZodMiniBigInt.init(inst, def);
});
// int64
// @__NO_SIDE_EFFECTS__
function int64(params) {
    return core._int64(exports.ZodMiniBigIntFormat, params);
}
// uint64
// @__NO_SIDE_EFFECTS__
function uint64(params) {
    return core._uint64(exports.ZodMiniBigIntFormat, params);
}
exports.ZodMiniSymbol = core.$constructor("ZodMiniSymbol", (inst, def) => {
    core.$ZodSymbol.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function symbol(params) {
    return core._symbol(exports.ZodMiniSymbol, params);
}
exports.ZodMiniUndefined = core.$constructor("ZodMiniUndefined", (inst, def) => {
    core.$ZodUndefined.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function _undefined(params) {
    return core._undefined(exports.ZodMiniUndefined, params);
}
exports.ZodMiniNull = core.$constructor("ZodMiniNull", (inst, def) => {
    core.$ZodNull.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function _null(params) {
    return core._null(exports.ZodMiniNull, params);
}
exports.ZodMiniAny = core.$constructor("ZodMiniAny", (inst, def) => {
    core.$ZodAny.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function any() {
    return core._any(exports.ZodMiniAny);
}
exports.ZodMiniUnknown = core.$constructor("ZodMiniUnknown", (inst, def) => {
    core.$ZodUnknown.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function unknown() {
    return core._unknown(exports.ZodMiniUnknown);
}
exports.ZodMiniNever = core.$constructor("ZodMiniNever", (inst, def) => {
    core.$ZodNever.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function never(params) {
    return core._never(exports.ZodMiniNever, params);
}
exports.ZodMiniVoid = core.$constructor("ZodMiniVoid", (inst, def) => {
    core.$ZodVoid.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function _void(params) {
    return core._void(exports.ZodMiniVoid, params);
}
exports.ZodMiniDate = core.$constructor("ZodMiniDate", (inst, def) => {
    core.$ZodDate.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function date(params) {
    return core._date(exports.ZodMiniDate, params);
}
exports.ZodMiniArray = core.$constructor("ZodMiniArray", (inst, def) => {
    core.$ZodArray.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function array(element, params) {
    return new exports.ZodMiniArray({
        type: "array",
        element: element,
        ...util.normalizeParams(params),
    });
}
// .keyof
// @__NO_SIDE_EFFECTS__
function keyof(schema) {
    const shape = schema._zod.def.shape;
    return _enum(Object.keys(shape));
}
exports.ZodMiniObject = core.$constructor("ZodMiniObject", (inst, def) => {
    core.$ZodObject.init(inst, def);
    exports.ZodMiniType.init(inst, def);
    util.defineLazy(inst, "shape", () => def.shape);
});
// @__NO_SIDE_EFFECTS__
function object(shape, params) {
    const def = {
        type: "object",
        shape: shape ?? {},
        ...util.normalizeParams(params),
    };
    return new exports.ZodMiniObject(def);
}
// strictObject
// @__NO_SIDE_EFFECTS__
function strictObject(shape, params) {
    return new exports.ZodMiniObject({
        type: "object",
        shape,
        catchall: never(),
        ...util.normalizeParams(params),
    });
}
// looseObject
// @__NO_SIDE_EFFECTS__
function looseObject(shape, params) {
    return new exports.ZodMiniObject({
        type: "object",
        shape,
        catchall: unknown(),
        ...util.normalizeParams(params),
    });
}
// object methods
// @__NO_SIDE_EFFECTS__
function extend(schema, shape) {
    return util.extend(schema, shape);
}
// @__NO_SIDE_EFFECTS__
function safeExtend(schema, shape) {
    return util.safeExtend(schema, shape);
}
// @__NO_SIDE_EFFECTS__
function merge(schema, shape) {
    return util.extend(schema, shape);
}
// @__NO_SIDE_EFFECTS__
function pick(schema, mask) {
    return util.pick(schema, mask);
}
// .omit
// @__NO_SIDE_EFFECTS__
function omit(schema, mask) {
    return util.omit(schema, mask);
}
// @__NO_SIDE_EFFECTS__
function partial(schema, mask) {
    return util.partial(exports.ZodMiniOptional, schema, mask);
}
// @__NO_SIDE_EFFECTS__
function required(schema, mask) {
    return util.required(exports.ZodMiniNonOptional, schema, mask);
}
// @__NO_SIDE_EFFECTS__
function catchall(inst, catchall) {
    return inst.clone({ ...inst._zod.def, catchall: catchall });
}
exports.ZodMiniUnion = core.$constructor("ZodMiniUnion", (inst, def) => {
    core.$ZodUnion.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function union(options, params) {
    return new exports.ZodMiniUnion({
        type: "union",
        options: options,
        ...util.normalizeParams(params),
    });
}
exports.ZodMiniXor = core.$constructor("ZodMiniXor", (inst, def) => {
    exports.ZodMiniUnion.init(inst, def);
    core.$ZodXor.init(inst, def);
});
/** Creates an exclusive union (XOR) where exactly one option must match.
 * Unlike regular unions that succeed when any option matches, xor fails if
 * zero or more than one option matches the input. */
function xor(options, params) {
    return new exports.ZodMiniXor({
        type: "union",
        options: options,
        inclusive: false,
        ...util.normalizeParams(params),
    });
}
exports.ZodMiniDiscriminatedUnion = core.$constructor("ZodMiniDiscriminatedUnion", (inst, def) => {
    core.$ZodDiscriminatedUnion.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function discriminatedUnion(discriminator, options, params) {
    return new exports.ZodMiniDiscriminatedUnion({
        type: "union",
        options,
        discriminator,
        ...util.normalizeParams(params),
    });
}
exports.ZodMiniIntersection = core.$constructor("ZodMiniIntersection", (inst, def) => {
    core.$ZodIntersection.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function intersection(left, right) {
    return new exports.ZodMiniIntersection({
        type: "intersection",
        left: left,
        right: right,
    });
}
exports.ZodMiniTuple = core.$constructor("ZodMiniTuple", (inst, def) => {
    core.$ZodTuple.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function tuple(items, _paramsOrRest, _params) {
    const hasRest = _paramsOrRest instanceof core.$ZodType;
    const params = hasRest ? _params : _paramsOrRest;
    const rest = hasRest ? _paramsOrRest : null;
    return new exports.ZodMiniTuple({
        type: "tuple",
        items: items,
        rest,
        ...util.normalizeParams(params),
    });
}
exports.ZodMiniRecord = core.$constructor("ZodMiniRecord", (inst, def) => {
    core.$ZodRecord.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function record(keyType, valueType, params) {
    return new exports.ZodMiniRecord({
        type: "record",
        keyType,
        valueType: valueType,
        ...util.normalizeParams(params),
    });
}
// @__NO_SIDE_EFFECTS__
function partialRecord(keyType, valueType, params) {
    const k = core.clone(keyType);
    k._zod.values = undefined;
    return new exports.ZodMiniRecord({
        type: "record",
        keyType: k,
        valueType: valueType,
        ...util.normalizeParams(params),
    });
}
function looseRecord(keyType, valueType, params) {
    return new exports.ZodMiniRecord({
        type: "record",
        keyType,
        valueType: valueType,
        mode: "loose",
        ...util.normalizeParams(params),
    });
}
exports.ZodMiniMap = core.$constructor("ZodMiniMap", (inst, def) => {
    core.$ZodMap.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function map(keyType, valueType, params) {
    return new exports.ZodMiniMap({
        type: "map",
        keyType: keyType,
        valueType: valueType,
        ...util.normalizeParams(params),
    });
}
exports.ZodMiniSet = core.$constructor("ZodMiniSet", (inst, def) => {
    core.$ZodSet.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function set(valueType, params) {
    return new exports.ZodMiniSet({
        type: "set",
        valueType: valueType,
        ...util.normalizeParams(params),
    });
}
exports.ZodMiniEnum = core.$constructor("ZodMiniEnum", (inst, def) => {
    core.$ZodEnum.init(inst, def);
    exports.ZodMiniType.init(inst, def);
    inst.options = Object.values(def.entries);
});
// @__NO_SIDE_EFFECTS__
function _enum(values, params) {
    const entries = Array.isArray(values) ? Object.fromEntries(values.map((v) => [v, v])) : values;
    return new exports.ZodMiniEnum({
        type: "enum",
        entries,
        ...util.normalizeParams(params),
    });
}
// @__NO_SIDE_EFFECTS__
/** @deprecated This API has been merged into `z.enum()`. Use `z.enum()` instead.
 *
 * ```ts
 * enum Colors { red, green, blue }
 * z.enum(Colors);
 * ```
 */
function nativeEnum(entries, params) {
    return new exports.ZodMiniEnum({
        type: "enum",
        entries,
        ...util.normalizeParams(params),
    });
}
exports.ZodMiniLiteral = core.$constructor("ZodMiniLiteral", (inst, def) => {
    core.$ZodLiteral.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function literal(value, params) {
    return new exports.ZodMiniLiteral({
        type: "literal",
        values: Array.isArray(value) ? value : [value],
        ...util.normalizeParams(params),
    });
}
exports.ZodMiniFile = core.$constructor("ZodMiniFile", (inst, def) => {
    core.$ZodFile.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function file(params) {
    return core._file(exports.ZodMiniFile, params);
}
exports.ZodMiniTransform = core.$constructor("ZodMiniTransform", (inst, def) => {
    core.$ZodTransform.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function transform(fn) {
    return new exports.ZodMiniTransform({
        type: "transform",
        transform: fn,
    });
}
exports.ZodMiniOptional = core.$constructor("ZodMiniOptional", (inst, def) => {
    core.$ZodOptional.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function optional(innerType) {
    return new exports.ZodMiniOptional({
        type: "optional",
        innerType: innerType,
    });
}
exports.ZodMiniExactOptional = core.$constructor("ZodMiniExactOptional", (inst, def) => {
    core.$ZodExactOptional.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function exactOptional(innerType) {
    return new exports.ZodMiniExactOptional({
        type: "optional",
        innerType: innerType,
    });
}
exports.ZodMiniNullable = core.$constructor("ZodMiniNullable", (inst, def) => {
    core.$ZodNullable.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function nullable(innerType) {
    return new exports.ZodMiniNullable({
        type: "nullable",
        innerType: innerType,
    });
}
// nullish
// @__NO_SIDE_EFFECTS__
function nullish(innerType) {
    return optional(nullable(innerType));
}
exports.ZodMiniDefault = core.$constructor("ZodMiniDefault", (inst, def) => {
    core.$ZodDefault.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function _default(innerType, defaultValue) {
    return new exports.ZodMiniDefault({
        type: "default",
        innerType: innerType,
        get defaultValue() {
            return typeof defaultValue === "function" ? defaultValue() : util.shallowClone(defaultValue);
        },
    });
}
exports.ZodMiniPrefault = core.$constructor("ZodMiniPrefault", (inst, def) => {
    core.$ZodPrefault.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function prefault(innerType, defaultValue) {
    return new exports.ZodMiniPrefault({
        type: "prefault",
        innerType: innerType,
        get defaultValue() {
            return typeof defaultValue === "function" ? defaultValue() : util.shallowClone(defaultValue);
        },
    });
}
exports.ZodMiniNonOptional = core.$constructor("ZodMiniNonOptional", (inst, def) => {
    core.$ZodNonOptional.init(inst, def);
    exports.ZodMiniType.init(inst, def);
});
// @__NO_SIDE_EFFECTS__
function nonoptional(innerType, params) {
    return new exports.ZodMiniNonOptional({
        type: "nonopt HčŃLčqMčōLČt$(LčHčXHčO Lc?Ičj∂CMč"E3“HkūpHčĄ$†   EćjHČĆ$ź   HČHčOHČt$xHČl$ DČEÖˇ~;HÉŃEč«HcAHk–8IV DąHčB(I#√H;B(ufDÖJA∂¬ADŇąHÉŃM+ŇuŐHčĆ$ź   Mč«IŃŗ3“ŤľÜ 3“H«GH   HČW0Hł.üáĘģBmTČW(ČW<HČG@ČWPHčD.`HČGXHčD.(HčH`HÖ…tIč$H9t	HčI(HÖ…uÚLčqHč◊IčőIčHč@ˇē= čūÖņt7ÉÝt2ÉÝuIč$Ť‘JýˇŽ"MčFMÖņu
čőŤś…  LčņHć¨ņ IčŐŤźpýˇIčNŤGCýˇE3ŘMČ^ÖŲtÉĢADůč∆ťÚ  ÉÕˇAčůEÖˇ~_AÉˇrYHćS8Lč
AćGˇHc»Ić…L; wH;¬s=Ač«%  Ä}ˇ»É»ĢˇņAč◊+–É∆;Ú|ýćBIč…ô+¬3“—ÝLcņIŃŗIÉŗÝŤĎÖ E3ŘHcőIčűI;Ō}
HčC8LČ»HőŽÓfDČ[Eč√HčGEÖˇéŚ   Lčt$(LćHHčĄ$ź   IčňLc»AćRˇÖ“ąģ   A;◊ćŁ   IcAÖņąū   A;Fćś   Hčs8MčŕE3“N9TřÝÖ—   E8Ą»   Lk–8MV IčB(H	č¬;’NČTřÝNŇE3ŘčŤHčĄ$ź   D8\»t'AÉÝ}!∑Ą$ė   D£ņr∑Cę–fČCHčĄ$ź   IčűAĄrtHčĒ$†   ÉgPĢDČ_<Č2D∆HőIÉŃI;ŌĆ/ˇˇˇć.AčňfČC,ÖŪxHčC8L9tVőHÉņ;Õ~ŮčG(ČCäG8ąCHčG0DČ_8HČC D9_<tTčGŽRHčL$ Hćí
 LčD$xMčD(IčŐMč ŤćnýˇAčŇť	  HčL$ Hći
 LčD$xMčD(IčŐMč ŤdnýˇťÍżˇˇAč√Út( ąCfDČ[ÚO@f/ŃÚĆ$†   rA∑ňŽhÚ") f/Ńr9Ú<) 3ņf/»rÚ\»f/»s
HĻ       ÄHčŃÚH,…H»Ťýˇ∑»Ž!HčĆ$†   łĢ  HŃť4f+»∑ŃfŃŗf»f…fČKHčOHŤť~ýˇčS(fČCč¬čOPļÍļū@"őHčL$pD–ČS(Hč”Ť≥śˇˇÄ{ čÝt
HčK ŤB@ýˇ∆C č«Hčú$Ä   HÉń0A_A^A]A\_^]√ŐŐŐŐ@SHÉž 3ŘLčŃÖ“xd;}`ģŤHc¬Hć@HčAHcLźIč@`Hk—8HčH Lč
IčAHÖņt-AųA   IčHhtIčQŽHč–IčALčņŤnQŁˇHÖņtHčŽHćť\ Hč√HÉń [√Lč‹IČ[IČKUVWATAUAVAWHĀžÄ   LčLčýHčILčÚHČL$hHč—Mč"IčoIčŐLČd$H∂ELk»pIčBHÉņL»IćC HČD$(IčGHČD$ ŤĹřˇˇ3ŘHčÝHÖņućGťf  fČ]Hč’«E(   fČ],ą]Hc0Ič$Dč∆Ťc‚ˇˇÖņt
Ľ   ť&  HćD$@E3…HČD$0IÉ»ˇ∑Ą$ō   Ič÷fČD$(IčŌHČ|$ ŤķˇˇčōÖņÖŖ  čD$@MčśIų‘Mč‘L#U LČT$PuÖņĄļ  E3ŪE3ŘL!l$`LČ\$XAćUÖņtMHćD$@Dč HČD$0IÉ»ˇ∑Ą$ō   Ič÷fČD$(IčŌHČ|$ ŤóýˇˇMč‹čōL#] LČ\$XAĒŇÖņÖZ  LčT$PHčőAčűHČL$xIÉ»ˇHÖ…~VHčD$hEćxHčWLč…LčT$`HÉ¬Hč@ LčōHcHk»8JčD(I#ńI;¬vI;ņLBņHÉ¬M+Ōu‹Lčľ$ņ   LčT$PLč\$XLČD$`IÉÝˇtfM;¬tOM;√tJHćD$@E3…HČD$0M∆∑Ą$ō   Ič÷fČD$(IčŌHČ|$ ŤŐÝˇˇčōLčT$PL9u uÉ|$@ ļ   DčÍDÚLč\$XHčL$xÖŘĄ0ˇˇˇÖŘuv∑¨$ō   EÖŪu9HćD$@E3…HČD$0Mč∆fČl$(Ič÷IčŌHČ|$ ŤgÝˇˇD9l$@čōAćEDūÖŘu2Žł   ÖŲu'Dč»Mč∆HćD$@Ič÷HČD$0IčŌfČl$(HČ|$ Ť%ÝˇˇčōLčd$HÉ8 t	HčO0ŤŌ<ýˇIč$Hč◊Ťk=ýˇč√Hčú$»   HĀńÄ   A_A^A]A\_^]√ŐŐHČ\$ UVWATAUAVAWHć¨$†żˇˇHĀž`  Hčś[ H3ńHČÖX  HčYWņMč»LČD$(E3ŲHČT$@Lč‚HČL$xHčQAčĢHČT$8HČ]ÄÖį  Lčz HcBLkņ8Öņ  Ö–  3ņLČ|$pHČÖŗ  M«∂CLk–pHčLČD$`LČT$HLčXLČ\$PCčDLČD$XM;ÝÉ  ∑Ķį  Čt$ ÖˇÖ  ł   fAÖGĄ‹  Mčw IčÜŤ  HÖCĄń  IcF«D$0   ILkŤ8HćÖŤ  Mn EąLČl$hA 3…MėHČMėE®HČE®Mčv M;űÉō  ł   fAÖFtIčF Ž,čD$XA9FÖó  HčHČEņHćEņHČU»∆E–,«E‘   LČuŗfČćŤ  Ič‘HČEźKčD(9HTHćMąt
MčŃŤWŻˇˇŽŤŮˇˇ3…čÝÖņuLčD$(HćMąIč‘Ť7ĢˇˇčÝ3…D∑•Ť  fEÖšĄ-  9L$0t8EčńfDČ•į  IŃŗHćēū  Hććł  ŤÉļ ∑Ķį  3…ČL$0Čt$ ť»   D∑∆Hćēł  IŃŗHćć(  fČĶ   ŤKļ D∑≠   3…ČL$ čŮfČćį  MÖŪĄĀ   AčńHćĶ0  čŔÖņtXLć•Ý  A∑T$∑NŤxýˇ∑D∑–A∑$ŤxýˇIčT$ÝHććį  HVÝD∑ņE∑ ŤĀ ˇˇ∑ÖŤ  Mćd$ˇ√;ō|Ī3…HÉ∆IÉŪuė∑Ķį  Hč]ÄČt$ Lčl$hHčT$8Lčd$@LčL$(LčT$HLč\$PIÉ∆8M;űsť;Ģˇˇ∑ŮfČćį  Čt$ LčD$`ł   WņfČC,DčŠHčC8LČ8«C(    fČKąKCÖˇuYLč|$xLćĶ¬  D∑ÓćwE;Ś}3A∑FĢHč”f∆IčŌfČCA∑fČCIčFŲHČŤ{ŖˇˇDśIÉ∆čÝÖņt»čt$ Lč|$pLčD$`Lč\$PLčT$HLčL$(Lčd$@HčT$8E3ŲHčL$xIÉ«8LČ|$pM;ÝāűŁˇˇč«HčćX  H3ŐŤŁd Hčú$ł  HĀń`  A_A^A]A\_^]√ŐHČ\$ UVWATAUAVAWHÉž HčyE3“Lč	HčŮEčŕLČT$hAčŕ«A, N  A∂A@MčAHk–pIčMćxLķIćhLč(HćGHHČG8AćBfČG0Ač¬fDČW,DČW(ČD$`I;ÔÉ  IćČX  HČL$pĀF,Ť  MčÚąGčUDŤa…ˇˇHČGäE<$
ŲōHčE MšM#„ÉxT tBLć]pM;Ŗs)Hč\$pMÖŲuAŲC<
tAčSDHčňŤ …ˇˇLūIÉ√pM;Ŗr‹Mč∆Ič‘HčőŤÝˇˇŽIč‘HčőŤŃŪˇˇE3“čōÖņuHčFD8PtMč∆Ič‘HčőŤōķˇˇčōE3“Lč\$hL_LČ\$hÖŘuE8UctÉŻeu6HćŘ ćK∑ŤŽYýˇLč\$hE3“AčŕčD$`HÉŇpHčL$pˇņČD$`I;Ôā
ˇˇˇHč◊IčÕŤ!ŕˇˇč√Hč\$xHÉń A_A^A]A\_^]√ŐŐLč‹fEČK MČCIČSIČKSUVWATAUAVAWHĀžė   E3šMčÝLč“HčŮEčŰAč‹fD9§$   tHčLčAŲA\@ÖŮ  E∑K ∑fČT$RÉķ?áŘ  ł   ČT$LDčņLČ§$Ä   DäōLČd$Xč DČ\$4I”ŗDčÍL+ņĻ  A∑ŃLČD$hf#ŃIč‘fųōHČT$pAčŐIčŁfŪČL$Hfų›fĀŇāfČl$2I;ōÉ:  ∑Ą$   ;»Ź'  Ö…~	I~HČ|$X;»}!IčGLč4–ł   LČī$ą   fDÖ»ÖŇ  ŽLčī$  LČī$ą   AčV(ļ‚
āŇ  A∑ŃĻÄ   f#ŃfČD$PtfEČfA∂FHk»pHčFDč|L3…DČ|$<DčŠEÖŪĄ^  DčťIćjAčńH£√ā	  HčM Ť§DŁˇ3…HčÝHÖņĄÚ   ä Z®żÖś   D9,Ö‹   ∑D$2HćNpLčL$XE3“DŅG0Ių—LČT$(Ač◊ČD$ Ť÷…ˇˇ3…LčÝHÖņĄü   DćIfD9Hu#A∑V,Ö“tMčF8HcŃM;<ņtA…; |ū; }q3…AŲGātbf9O0|\Hč>HčU HčŌŤ*DŁˇHčūHÖņuHčHčpIčų@   t
HčPLč@ŽHčPLč@HčŌŤEFŁˇ3…HÖņtHčHčŤĮbýˇ3…ÖņuLęŽŽ3…Dč|$<Hčī$ŗ   ł   DŗLŤHÉŇ D;d$LĆ»ĢˇˇAčV(Dč\$4LčĒ$Ť   D∑Ć$Ý   ∑l$2ļ‚āc  ļ‚sĺ   fČL$D∑∆LčťŽJMčn MÖŪĄF  AŲEdÖ;  A∑E^fČD$DA∑E`A8Mbtļ‚r
ĺ   DäřŽDäŔĺ   DČ\$4∑ņDäŠąL$9čýąL$8DČd$@ČD$`ÖņĄŗ  E∑FA;Ý}|A∑F.;Ý|sMčN8Hc«IčŃ∑AfÖŤt/D∑Ć$Ý   ĻÄ  fÖŃĻ    A∂√D»DäŔ3…DČ\$4ćqť~  Ĺ   @ĄŇtLčćWŽHc¬IčŃL9t’A;–|Ū3…Ž3…@äŮŽĹ   MÖŪĄ  IčEHcŌDŅ<HIčE8äIčE@"’ąT$0ŅHDD;ýuAÉŌˇ3…EĄŘt7EÖˇx!A;Ý|IčEHčHIc«HŃŗäDŲō…D"Ŕ3…AÉˇĢE∂ŘDDŔDČ\$4čť@ĄŲĄĘ  ;l$Lćé  čŇH£√āw  LčĄ$Ť   HcŇLčŗHČD$xIŃšKčLŤóAŁˇ3…Hč–HÖņĄH  äĄ$Ý   $ņŲōņ@"∆ąD$:@äūAÉˇˇ|9Ä:¶t	Ä:®Ö  čL$<9J,Ö  ŅB03…A;«Ö  ŽNAÉŌˇąL$0ťˇˇˇIčEPHcŌHŃŠHčLŤAŁˇHč LčņŤAŁˇDčL$<Hč–3…ŤžĘŁˇ3…ÖņÖĻ   LčĄ$Ť   AÉˇˇtEHčĄ$ŗ   KčTHč0HčőŤ AŁˇHč»HÖņuHčHčHIčU@Hč	@ät$:Hc«Hč¬Ť•_ýˇ3…Öņufĺ   f9L$Ptć>fAČFD∑Ć$Ý   @ä÷LčD$xLčĒ$Ť   AŲŃ@Ö   IčņHŃŗBäL3ņ@"ő8D$9Ą◊   äD$02D$8:ŃĒ¬ťŪ   3…ˇŇ@ĄŲÖhĢˇˇDč\$4Dčd$@LčĒ$Ť   ĺ   D∑Ć$Ý   Öˇt	∑D$D;Ý}DäŔEĄšE∂ŘDEřDČ\$4Dčl$LEĄŘĄi  Lč§$Ä   čýMfLČ§$Ä   EÖŪĄJ  Lčī$ŗ   Ał   č«H£√ā  LcˇIč«HŃŗJčlHÖŪĄĶ   IćéX  Hč’ŤŅˇˇHčū3…ťü   äŃ2D$0ąD$8tLčú$  HcD$HIčHęŃIČ@ąt$93…Ą“t_IčņHŃŗBŲDtA∑F;ÝuAļn(Žä—Dčd$@Dč\$4Ą“Ąˇˇˇ∑l$2AÉˇˇE∂šDDśLę√DČd$@Ģ;|$`ćūĢˇˇťŁˇˇDč\$4Dčd$@ťŐĢˇˇHčŮHÖŲuE3ņćVHčÕŤ`Łˇ3…ÖņtIčńAł   Hų–HÖ∆uLęŻŽAł   LčĒ$Ť   AÝA;żĆŕĢˇˇLčī$ą   Dč\$4D∑Ć$Ý   LčD$hE3ščL$HHčī$ŗ   ∑l$2Hč|$XLčľ$ū   ł   HD$p»HčT$pČL$HEĄŘÖ‘ÝˇˇŽE8ftłÄ   DĄ»IDōI;ōuäD$RŽ0EĄŘuCAćMˇÖ…~!Ał   Ič–H”‚I+–Hč¬H#√H;¬tA+»Ö…Ś2ņHĀńė   A_A^A]A\_^][√äŃŽ
